File: src/javascript/algorithms/dynamicProgramming/regularExpressionMatching/regularExpressionMatching.js
// Approach 1: Top-down memoization
function isMatch(s, p) {
    const m = s.length, n = p.length;
    const memo = new Map();

    function dfs(i, j) {
        const key = `${i},${j}`;
        if (memo.has(key)) return memo.get(key);
        if (j === n) {
            memo.set(key, i === m);
            return i === m;
        }

        const match = i < m && (p[j] === s[i] || p[j] === '.');

        let result;
        if (j + 1 < n && p[j + 1] === '*') {
            result = dfs(i, j + 2) || (match && dfs(i + 1, j));
        } else {
            result = match && dfs(i + 1, j + 1);
        }

        memo.set(key, result);
        return result;
    }

    return dfs(0, 0);
}

// Approach 2: Bottom-up DP
function isMatch1(s, p) {
    const m = s.length, n = p.length;
    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(false));
    dp[m][n] = true;

    for (let i = m; i >= 0; i--) {
        for (let j = n - 1; j >= 0; j--) {
            const match = i < m && (p[j] === s[i] || p[j] === '.');
            if (j + 1 < n && p[j + 1] === '*') {
                dp[i][j] = dp[i][j + 2] || (match && dp[i + 1][j]);
            } else {
                dp[i][j] = match && dp[i + 1][j + 1];
            }
        }
    }

    return dp[0][0];
}

function runTests() {
    const tests = [
        { s: 'aa',  p: 'a',     expected: false },
        { s: 'aa',  p: 'a*',    expected: true  },
        { s: 'ab',  p: '.*',    expected: true  },
        { s: 'aab', p: 'c*a*b', expected: true  },
        { s: 'a',   p: '.',     expected: true  },
    ];

    tests.forEach(({ s, p, expected }, i) => {
        const r1 = isMatch(s, p);
        const r2 = isMatch1(s, p);
        console.log(`${r1 === expected ? 'PASS' : 'FAIL'} test ${i + 1} isMatch:  result=${r1} expected=${expected}`);
        console.log(`${r2 === expected ? 'PASS' : 'FAIL'} test ${i + 1} isMatch1: result=${r2} expected=${expected}`);
    });
}

runTests();

File: src/python/algorithms/dynamicProgramming/regularExpressionMatching/regular_expression_matching.py
def is_match(s, p):
    """Top-down memoization"""
    m, n = len(s), len(p)
    memo = {}

    def dfs(i, j):
        if (i, j) in memo:
            return memo[(i, j)]
        if j == n:
            memo[(i, j)] = (i == m)
            return memo[(i, j)]

        match = i < m and (p[j] == s[i] or p[j] == '.')

        if j + 1 < n and p[j + 1] == '*':
            result = dfs(i, j + 2) or (match and dfs(i + 1, j))
        else:
            result = match and dfs(i + 1, j + 1)

        memo[(i, j)] = result
        return result

    return dfs(0, 0)


def is_match1(s, p):
    """Bottom-up DP"""
    m, n = len(s), len(p)
    dp = [[False] * (n + 1) for _ in range(m + 1)]
    dp[m][n] = True

    for i in range(m, -1, -1):
        for j in range(n - 1, -1, -1):
            match = i < m and (p[j] == s[i] or p[j] == '.')
            if j + 1 < n and p[j + 1] == '*':
                dp[i][j] = dp[i][j + 2] or (match and dp[i + 1][j])
            else:
                dp[i][j] = match and dp[i + 1][j + 1]

    return dp[0][0]


if __name__ == "__main__":
    tests = [
        ("aa",  "a",     False),
        ("aa",  "a*",    True),
        ("ab",  ".*",    True),
        ("aab", "c*a*b", True),
        ("a",   ".",     True),
    ]

    for i, (s, p, expected) in enumerate(tests, 1):
        r1 = is_match(s, p)
        r2 = is_match1(s, p)
        print(f"{'PASS' if r1 == expected else 'FAIL'} test {i} is_match:  result={r1} expected={expected}")
        print(f"{'PASS' if r2 == expected else 'FAIL'} test {i} is_match1: result={r2} expected={expected}")
